### PR TITLE
Issue 71 refactor simulation app

### LINKED ISSUE #71: Refactor simulation app
- [x] Replace the demo app with the improved erlotinib PDSimulationController app.

### PR DESCRIPTION
**Pull request by DavAug**
Closes #71 .

### CODE REVIEW

File: pkpdapp/pkpdapp/settings.py [hunk 1]
```diff
@@ -164,16 +164,62 @@
...
+
+    # Flag controlling local serving of assets
+    "serve_locally": False,
+}
+
+# Staticfiles finders for locating dash app assets and related files
+
+STATICFILES_FINDERS = [
+
+    'django.contrib.staticfiles.finders.FileSystemFinder',
```
Comment (ben18785) [inline pkpdapp/pkpdapp/settings.py:198]: Can we make these alphabetical?

File: pkpdapp/pkpdapp/settings.py [hunk 2]
```diff
@@ -164,16 +164,62 @@
...
+]
+
 # Plotly components containing static content that should
 # be handled by the Django staticfiles infrastructure
 
 PLOTLY_COMPONENTS = [
+
+    # Common components
     'dash_core_components',
     'dash_html_components',
```
Comment (ben18785) [inline pkpdapp/pkpdapp/settings.py:213]: Same -- alphabetical too please

File: pkpdapp/simulate/dash_apps/simulation.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        1. Pharmacokinetic input
+        2. Initial values (of states)
+        3. Parameters
+        """
+        parameters = self._model.parameters()
+        # Add one slider for each parameter
+        for parameter in parameters:
+            self._sliders.add_slider(slider_id=parameter)
+
+        # Split parameters into initial values, and parameters
```
Comment (ben18785) [inline pkpdapp/simulate/dash_apps/simulation.py:92]: "Split parameters into initial values and ODE parameters"

**Review by ben18785 — CHANGES_REQUESTED** [review]
Thanks @DavAug. As far as I can tell given my base experience, looks good. Only a couple of minor things. I'll add Martin as a reviewer though.

**Review by martinjrobins — CHANGES_REQUESTED** [review]
thanks @DavAug, this looks great. just need to up the test coverage I think

File: pkpdapp/pkpdapp/settings.py (hunk 1 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/settings.py:198]: I'm easy about it. You could group by functionality (with comments to label groups of functions) and do alphabetical within them but really not an important thing!

File: pkpdapp/pkpdapp/settings.py (hunk 2 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/settings.py:213]: On reflection, think you're right here!

**Comment by DavAug** [comment]
I have written tests for the app, but we need to adapt the run-tests.py to run the tests for each of the sub apps.

**Comment by DavAug** [comment]
Will do that later this week.

**Comment by DavAug** [comment]
Hi @martinjrobins , 
The unit tests seem to fail on the database models now (I haven't touched them though!). Is it possible that your recent PR broke something?

This branch is up to date with the master now.

**Comment by martinjrobins** [comment]
looks like  a bug in django-test-migrations? I don't get it on my machine, do you?

**Comment by DavAug** [comment]
> looks like a bug in django-test-migrations? I don't get it on my machine, do you?

When I run the migrations (without making them again) I get already an error here:
```
python manage.py migrate

Operations to perform:
  Apply all migrations: admin, auth, contenttypes, django_plotly_dash, pkpdapp, sessions
Running migrations:
  Applying pkpdapp.0006_initial_pkpd_models... OK
  Applying pkpdapp.0005_initial_datasets...Traceback (most recent call last):
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.OperationalError: table pkpdapp_dataset has no column named datetime

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 245, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/pkpdapp/pkpdapp/migrations/0005_initial_datasets.py", line 53, in load_datasets
    dataset.save()
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/base.py", line 746, in save
    force_update=force_update, update_fields=update_fields)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/base.py", line 784, in save_base
    force_update, using, update_fields,
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/base.py", line 887, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/base.py", line 926, in _do_insert
    using=using, raw=raw,
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/query.py", line 1204, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 1392, in execute_sql
    cursor.execute(sql, params)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.OperationalError: table pkpdapp_dataset has no column named datetime

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 29, in <module>
    main()
  File "manage.py", line 25, in main
    execute_from_command_line(sys.argv)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/base.py", line 328, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/core/management/commands/migrate.py", line 233, in handle
    fake_initial=fake_initial,
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/migrations/executor.py", line 247, in apply_migration
    migration_recorded = True
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/sqlite3/schema.py", line 35, in __exit__
    self.connection.check_constraints()
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/sqlite3/base.py", line 307, in check_constraints
    violations = self.cursor().execute('PRAGMA foreign_key_check').fetchall()
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/utils.py", line 80, in _execute
    self.db.validate_no_broken_transaction()
  File "/Users/david/Workspace/sabs-r3/pkpdapp/venv/lib/python3.7/site-packages/django/db/backends/base/base.py", line 449, in validate_no_broken_transaction
    "An error occurred in the current transaction. You can't "
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

**Comment by DavAug** [comment]
Also makemigrations crashes

**Comment by martinjrobins** [comment]
try to do the migration on a clean database (i.e. delete the sqlite file and then run `python manage.py. migrate`). Remember your database doesn't care which git branch you are in. So these steps will fail:

1. run migrations A
2. change to branch which modifies migration A and also adds migration B
3.  run migrations again, which tries to apply B (since A has already been applied). B fails cause it assumes the updated A has been run

`python manage.py makemigrations` doesn't do anything to your database, it just looks at your database and looks at your models, and works out how to change your database to fit the current models and generates a `.py` migration file to do these changes.

**Comment by DavAug** [comment]
> try to do the migration on a clean database (i.e. delete the sqlite file and then run `python manage.py. migrate`). Remember your database doesn't care which git branch you are in. So these steps will fail:
> 
> 1. run migrations A
> 2. change to branch which modifies migration A and also adds migration B
> 3. run migrations again, which tries to apply B (since A has already been applied). B fails cause it assumes the updated A has been run
> 
> `python manage.py makemigrations` doesn't do anything to your database, it just looks at your database and looks at your models, and works out how to change your database to fit the current models and generates a `.py` migration file to do these changes.

Ahhh yes you are absolutely right! This time it works, and the test were successful. Any thoughts on what we do with the actions though? It seems weird that it runs on our machines and not in the containers.

**Comment by martinjrobins** [comment]
I'm going to hazard a guess that the containers get a newer version of some library, which causes the bug. Perhaps pin the dependencies to the specific versions you have installed now?

**Review by martinjrobins — APPROVED** [review]
I'm happy to merge when you are ready @DavAug. Feel free to either fix those errors in this PR (I think if you pin the dependency versions to what you are using on your machine it should work), or we can do it in a separate PR, given that it is not related to this issue.

**Comment by DavAug** [comment]
> I'm happy to merge when you are ready @DavAug. Feel free to either fix those errors in this PR (I think if you pin the dependency versions to what you are using on your machine it should work), or we can do it in a separate PR, given that it is not related to this issue.

Let me try to pin them here, and then we can megre :)

**Comment by DavAug** [comment]
@martinjrobins good shout!

**Comment by DavAug** [comment]
@martinjrobins Do you have an idea why codecov is not updating the patch not the project coverage?

**Comment by martinjrobins** [comment]
> @martinjrobins Do you have an idea why codecov is not updating the patch nor the project coverage?

Looks like the coverage workflow is erroring with: "Error: No coverage report found". Don't know why github is giving it a green tick then!

**Comment by DavAug** [comment]
> > @martinjrobins Do you have an idea why codecov is not updating the patch nor the project coverage?
> 
> Looks like the coverage workflow is erroring with: "Error: No coverage report found". Don't know why github is giving it a green tick then!

I am not sure why codecov can't find older reports. Is it possible that it has no copmarison, because I changed how the tests are executed, i.e. before we had 
```
python pkpdapp/manage.py test pkpdapp
```
but in order to test all apps with one command we need to call manage.py test in the same directory as the apps, i.e.
```
cd pkpdapp && python manage.py test
```

**Pull request merged**